Draft docs for using docker testclusters. (#20397)
hashicorp/vault: README.md

you what to set, so it is not documented here.

For more information on Vault Enterprise features, visit the [Vault Enterprise site](https://www.hashicorp.com/products/vault/?utm_source=github&utm_medium=referral&utm_campaign=github-vault-enterprise).

### Docker-based Tests

We have created an experimental new testing mechanism inspired by NewTestCluster.
A minimalist example of how to use it:

```go
import (
  "testing"
  "github.com/hashicorp/vault/sdk/helper/testcluster/docker"
)

func Test_Something_With_Docker(t *testing.T) {
  opts := &docker.DockerClusterOptions{
    ImageRepo: "hashicorp/vault",
    ImageTag:    "latest",
  }
  cluster := docker.NewTestDockerCluster(t, opts)
  defer cluster.Cleanup()
}
```

A more realistic example follows.  In this case, the test should be run

> with VAULT_BINARY pointing at a local Linux amd64 vault binary on disk

Author: ```suggestion
with VAULT_BINARY pointing at a local Linux vault binary on disk
```

you what to set, so it is not documented here.

For more information on Vault Enterprise features, visit the [Vault Enterprise site](https://www.hashicorp.com/products/vault/?utm_source=github&utm_medium=referral&utm_campaign=github-vault-enterprise).

### Docker-based Tests

> We have created an experimental new testing mechanism inspired by NewTestCluster.

Reviewer: Is this still "experimental"? We're beginning to build test scenarios on this that are not easy to build any other way and I'd consider those to be first class 🤔 . I guess it's "experimental" in the sense of "no external API guarantees" though?

Author: Yeah, that was my only reason for using that word.

A minimalist example of how to use it:

```go
import (
  "testing"
  "github.com/hashicorp/vault/sdk/helper/testcluster/docker"
)

func Test_Something_With_Docker(t *testing.T) {
  opts := &docker.DockerClusterOptions{
    ImageRepo: "hashicorp/vault",
    ImageTag:    "latest",
  }
  cluster := docker.NewTestDockerCluster(t, opts)
  defer cluster.Cleanup()
}
```

Here is a more realistic example.  This test should be run with VAULT_BINARY pointing
at a local Linux vault binary on disk that you wish to run inside the docker containers. 
If it's an Enterprise binary, you'll also need to set VAULT_LICENSE_CI, or the
DockerClusterOptions field VaultLicnese. 

Optionally you can set COMMIT_SHA, which will be appended to the image name we
build as a debugging convenience.

```go
func Test_Custom_Build_With_Docker(t *testing.T) {
  binary := os.Getenv("VAULT_BINARY")
  if binary == "" {
    t.Skip("only running docker test when $VAULT_BINARY present")

>   }

Author: Ah, right, I forgot we'd added DefaultOptions, good call.

  opts := &docker.DockerClusterOptions{
    ImageRepo: "hashicorp/vault",
    // We're replacing the binary anyway, so we're not too particular about
    // the docker image version tag.
    ImageTag:    "latest",
    VaultBinary: binary,
    ClusterOptions: testcluster.ClusterOptions{
      VaultNodeConfig: &testcluster.VaultNodeConfig{
        LogLevel: "TRACE",
      },
    },
  }
  cluster := docker.NewTestDockerCluster(t, opts)
  defer cluster.Cleanup()
}
```

There are a variety of helpers in the `github.com/hashicorp/vault/sdk/helper/testcluster`
package, e.g. these tests below will create a pair of 3-node clusters and link them using

> PR or DR replication respectively, and fail if the replication state doesn't become healthy.

Author: Good point, will add.

func Test_Something_With_Docker(t *testing.T) {
  opts := &docker.DockerClusterOptions{

>     ImageRepo: "hashicorp/vault", // or "hashicorp/vault-enterprise"

Reviewer: Do we need this comment if we have a whole separate example below for enterprise which includes the licensing config?

Author: I'm kind of trying to cater to two different audiences here, external and internal.  For internal, we're generally going to want to be testing using locally built binaries (on dev machine or in CI), and so the repo is irrelevant.  For external, they're going to be testing their integration with vault rather than testing their own changes to vault, so it may be easier/faster for them to just use an upstream docker image.

    ImageTag:    "latest",
  }
  cluster := docker.NewTestDockerCluster(t, opts)
  defer cluster.Cleanup()
  
  client := cluster.Nodes()[0].APIClient()
  _, err := client.Logical().Read("sys/storage/raft/configuration")
  if err != nil {
    t.Fatal(err)
  }
}
```

Or for Enterprise:

```go
import (
  "testing"
  "github.com/hashicorp/vault/sdk/helper/testcluster/docker"
)

func Test_Something_With_Docker(t *testing.T) {
  opts := &docker.DockerClusterOptions{
    ImageRepo: "hashicorp/vault-enterprise",
    ImageTag:  "latest",
	VaultLicense: licenseString, // not a path, the actual license bytes
  }
  cluster := docker.NewTestDockerCluster(t, opts)
  defer cluster.Cleanup()
}
```

Here is a more realistic example of how we use it in practice.  DefaultOptions uses 
`hashicorp/vault`:`latest` as the repo and tag, but it also looks at the environment
variable VAULT_BINARY. If populated, it will copy the local file referenced by

> VAULT_BINARY into the container. This is useful when there is no published image
> containing the vault binary you want to test.

Reviewer: ```suggestion
VAULT_BINARY into the container. This is useful when testing local changes.
```
Very much optional. Your version is more technically accurate but my guess is this says more directly what readers would need to infer from the sentence otherwise.

Instead of setting the VaultLicense option, you can set the VAULT_LICENSE_CI environment
variable, which is better than committing a license to version control.

Optionally you can set COMMIT_SHA, which will be appended to the image name we
build as a debugging convenience.

```go
func Test_Custom_Build_With_Docker(t *testing.T) {
  opts := docker.DefaultOptions(t)
  cluster := docker.NewTestDockerCluster(t, opts)
  defer cluster.Cleanup()
}
```

There are a variety of helpers in the `github.com/hashicorp/vault/sdk/helper/testcluster`
package, e.g. these tests below will create a pair of 3-node clusters and link them using
PR or DR replication respectively, and fail if the replication state doesn't become healthy
before the passed context expires.

Again, as written, these depend on having a Vault Enterprise binary locally and the env
var VAULT_BINARY set to point to it, as well as having VAULT_LICENSE_CI set.

```go
func TestStandardPerfReplication_Docker(t *testing.T) {
  opts := docker.DefaultOptions(t)
  r, err := docker.NewReplicationSetDocker(t, opts)
  if err != nil {
      t.Fatal(err)
  }
  defer r.Cleanup()

>   ctx, cancel := context.WithTimeout(context.Background(), time.Minute)

Reviewer: nit: is a minute long enough to not be too flaky in our experience? People will be able to figure that out of course but if these examples are reasonable starting points to copy directly including sensible timing info that seems more useful.

I don't know that it isn't reasonable as it is, but I've seen a few of the more complex examples have timeouts of multiple minutes to be more resilient in slow CI.
